### PR TITLE
[WIP] new db layout, better recurring events support

### LINKED ISSUE #118: new database layout
was `blacklist VTODOs in khal's db`

```
 khal tries to parse them every time the vdir updates, which is kind of unperformant. Could we blacklist these items in the DB until they actually update?
```

I'm currently rethinking the db layout to deal better with complex recurrence patterns (see branch for the current iteration `rrrule_exceptions`), this won't happen before that is done.

### PR DESCRIPTION
**Pull request by geier**
this is mainly a todo list:

- [x]  support empty ics (i.e. no vevents inside, but VTODOs)
- [x] test for empty ics
- [x] EXDATE
- [x] RDATE with DATE and DATE-TIME values
- [x] deal with RDATE with PERIOD
    -  implementation currently not possible, [need to support this in icalendar first](https://github.com/collective/icalendar/issues/156)
  - [x] print warning instead
- [x] RECURRENCE-ID (see #95 for example)
- [x] RECURRENCE-ID;RANGE=THISANDFUTURE:19980401T133000Z
- [x] explicit non-suppot for RECURRENCE-ID;RANGE=THISANDPRIOR
- [x] support for events with same uid in different calendars (while not needed for standard conformity this is probably nice to have)
- [x] check if more than one RANGE=THISANDFUTURE is allowed and if so implement, otherwise at least warn if this happes

once this is ready it should close #118 and  #68

### CODE REVIEW

**Comment by untitaker** [comment]
I pushed a commit that skips the locale test if the DE-locales are not generated. Just FYI, so please don't ``push -f`` :)

File: tests/backend_test.py [hunk 1]
```diff
@@ -38,4 +52,155 @@ def test_new_db_version():
...
+def test_event_rrule_this_and_future():
+    dbi = backend.SQLiteDb('home', ':memory:', locale=locale)
+    dbi.update(event_rrule_this_and_future, href='12345.ics', etag='abcd')
+    assert dbi.list() == [('12345.ics', 'abcd')]
+    events = dbi.get_time_range(datetime(2014, 4, 30, 0, 0), datetime(2014, 9, 26, 0, 0))
+    events.sort(key=lambda x: x.start)
+    assert len(events) == 6
+
+    assert events[0].start == berlin.localize(datetime(2014, 6, 30, 7, 0))
+    assert events[1].start == berlin.localize(datetime(2014, 7, 7, 9, 0))
```
Comment (untitaker) [inline tests/backend_test.py]: This currently fails. Instead of the expected value, it gives me:

    berlin.localize(datetime(2014, 7, 7, 7, 0))

Comment (untitaker) [inline tests/backend_test.py]: None of the recurring events should start at 9AM?

Comment (geier) [inline tests/backend_test.py]: this is not implemented yet...

**Comment by geier** [comment]
thanks for properly fixing that locale test

File: tests/backend_test.py (hunk 1 above)
Comment (untitaker) [inline tests/backend_test.py]: I don't understand... the output of the tested code looks correct to me. Note that I am not well-versed with dates and times though.

Comment (geier) [inline tests/backend_test.py]:
```RECURRENCE-ID;RANGE=THISANDFUTURE:20140707T050000Z
DTSTART;TZID=Europe/Berlin:20140707T090000```

means all events after 20140707T050000Z should start at 9:00.

Comment (untitaker) [inline tests/backend_test.py]: Oh, I missed those lines... That's horrifying to see, that the order of DTSTART/RECURRENCE-ID items seems to matter?

Comment (geier) [inline tests/backend_test.py]: no, the order doesn't matter, but all events with the same UID are updated. Actually the order can be arbitrary anywhere (that's why the events are not directly inserted into the db from the icalendar object, but sorted before).

Comment (untitaker) [inline tests/backend_test.py]: Screw me... I oversaw that these are two separate events.

**Comment by untitaker** [comment]
BTW how does this behave if there is a VTODO? Does it save the VTODO in the db, or create a stub entry for it?

**Comment by geier** [comment]
atm VTODOS **should** be inserted into the `events` table, but not in
recurring date(time) tables. This could of course be changed to include
e.g. their due date or reminder time.

Quoting Markus Unterwaditzer (2015-01-29 12:17:40)

**Comment by untitaker** [comment]
I added a new testcase and changed some logging messages. The new behavior is:

* Warn if a file can't be parsed
* Just skip silently if no vevent could be found

**Pull request merged**